raiden/storage/utils.py

```diff
@@ -58,7 +58,7 @@ def __getattr__(self, item):
 CREATE TABLE IF NOT EXISTS state_changes (
     identifier ULID PRIMARY KEY NOT NULL,
     data JSON,
-    timestamp TIMESTAMP DEFAULT CURRENT_TIMESTAMP NOT NULL
+    timestamp TIMESTAMP DEFAULT(STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW')) NOT NULL
```

Comment:
This is not valid ISO8601

Reply:
> In formats 5 through 7, the "T" is a literal character separating the date and the time, as required by ISO-8601

https://www.sqlite.org/lang_datefunc.html

Reply:
+1 you are missing the seconds there

Reply:
Just make sure that all the examples in the docs and what you implement here follow ISO8601:  https://raiden-network.readthedocs.io/en/latest/rest_api.html#get--api-v1-payments-(token_address)-(target_address)

Reply:
Yeap, missing the T to separate date from time. There was some issue with the sqlite library itself when it was trying to actually parse the string with the `T` in there. I will try to repro it and paste it here.

Reply:
@hackaugusto: got to repro the error when placing `T` in the strftime string. apparently sqlite does not follow ISO here and expects a space to separate date from time.

```$ pytest raiden/tests/unit/ --lf
============================================================================================================================= test session starts =============================================================================================================================
platform linux -- Python 3.7.3, pytest-4.6.3, py-1.8.0, pluggy-0.12.0
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/home/raphael/code/brainbot/raiden/client/.hypothesis/examples')
rootdir: /home/raphael/code/brainbot/raiden/client, inifile: setup.cfg
plugins: xdist-1.28.0, forked-1.0.2, random-0.2, select-0.1.2, hypothesis-4.24.3
collected 25 items / 23 deselected / 2 selected                                                                                                                                                                                                                               
run-last-failure: rerun previous 2 failures (skipped 1134 files)

raiden/tests/unit/test_sqlite.py F                                                                                                                                                                                                                                      [ 50%]
raiden/tests/unit/test_wal.py F                                                                                                                                                                                                                                         [100%]

================================================================================================================================== FAILURES ===================================================================================================================================
________________________________________________________________________________________________________________________________ test_log_run _________________________________________________________________________________________________________________________________

    def test_log_run():
        with patch("raiden.storage.sqlite.get_system_spec") as get_speck_mock:
            get_speck_mock.return_value = dict(raiden="1.2.3")
            serializer = JSONSerializer()
            store = SerializedSQLiteStorage(":memory:", serializer)
            store.log_run()
        cursor = store.database.conn.cursor()
>       cursor.execute("SELECT started_at, raiden_version FROM runs")

raiden/tests/unit/test_sqlite.py:372: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

val = b'2019-08-14T10:38:26.447'

    def convert_timestamp(val):
>       datepart, timepart = val.split(b" ")
E       ValueError: not enough values to unpack (expected 2, got 1)

/usr/lib/python3.7/sqlite3/dbapi2.py:67: ValueError
___________________________________________________________________________________________________________________________ test_write_read_events ____________________________________________________________________________________________________________________________

    def test_write_read_events():
        wal = new_wal(state_transition_noop)
    
        event = EventPaymentSentFailed(
            factories.make_token_network_registry_address(),
            factories.make_address(),
            1,
            factories.make_address(),
            "whatever",
        )
    
        with pytest.raises(sqlite3.IntegrityError):
            unexisting_state_change_id = random.getrandbits(16 * 8).to_bytes(16, "big")
            wal.storage.write_events([(unexisting_state_change_id, event)])
    
        previous_events = wal.storage.get_events_with_timestamps()
    
        state_change_ids = wal.storage.write_state_changes([StateChange()])
        wal.storage.write_events([(state_change_ids[0], event)])
    
>       new_events = wal.storage.get_events_with_timestamps()

raiden/tests/unit/test_wal.py:164: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
raiden/storage/sqlite.py:859: in get_events_with_timestamps
    limit=limit, offset=offset, filters=filters, logical_and=logical_and
raiden/storage/sqlite.py:662: in get_events_with_timestamps
    limit=limit, offset=offset, filters=filters, logical_and=logical_and
raiden/storage/sqlite.py:594: in _query_events
    logical_and=logical_and,
raiden/storage/sqlite.py:475: in _form_and_execute_json_query
    cursor.execute(query, args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

val = b'2019-08-14T10:38:26.526'

    def convert_timestamp(val):
>       datepart, timepart = val.split(b" ")
E       ValueError: not enough values to unpack (expected 2, got 1)

/usr/lib/python3.7/sqlite3/dbapi2.py:67: ValueError
=================================================================================================================== 2 failed, 23 deselected in 0.55 seconds```

@LefterisJP, not really. sqlite strftime is different from python. `%f` represents *seconds and miliseconds*

Reply:
> @LefterisJP, not really. sqlite strftime is different from python. `%f` represents *seconds and miliseconds*

Interesting. I did not know that. Which is why having tests to make sure all of this works as expected is needed.

Reply:
I added an integration test now. The "T" is still missing on the timestamp string to actually be considered a proper ISO8601 string. However I tested with both Python and Javascript and they both managed to parse the string with a blank space.

To actually get a fully-compliant string, the only sane-ish approach that I can think now is to to turn the timestamp on sqlite to regular string fields and some marshmallow massaging of the fields.

If we are okay with having a working-but-not-really-ISO timestamp, then I will also write an unit test and it should be okay to merge.

Reply:
Can you add a unit test as well? Seems like there should be a quick way of testing this without starting the whole integration test stuff